File: Juan/P2/java/Napakalaki/src/napakalaki/CardDealer.java

package napakalaki;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;

/**
 *
 * @author juane
 * @version 2.1
 */
public class CardDealer {
    /**
     * Atributos referenciales
     */
    private ArrayList <Monster> unusedMonsters;
    private ArrayList <Monster> usedMonsters;
    private ArrayList <Treasure> unusedTreasures;
    private ArrayList <Treasure> usedTreasures;
    
    private static final CardDealer instance = null;
    
    /** Constructor ****/
    
    /**
     * Constructor CardDealer()
     * visibilidad privada
     */
    private CardDealer(){
        
    }
     
    /**
     * Método initTreasureCardDeck()
     */
    private void initTreasureCardDeck(){
        
         unusedTreasures.add(new Treasure("¡Si mi amo!",4,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("Botas de investigacion",3,TreasureKind.SHOES));
         
         unusedTreasures.add(new Treasure("Capucha de Cthulhu",3,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("A prueba de babas",2,TreasureKind.ARMOR));
         
         unusedTreasures.add(new Treasure("Botas de lluvia ácida",1,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Casco minero",2,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("Ametralladora Thompson",4,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Camiseta de la UGR",1,TreasureKind.ARMOR));
         
         unusedTreasures.add(new Treasure("Clavo de rail ferroviario",3,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Cuchillo de sushi arcano",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Fez alópodo",3,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("Hacha prehistórica",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("El aparato del Pr. Tesla",4,TreasureKind.ARMOR));
         
         unusedTreasures.add(new Treasure("Gaita",4,TreasureKind.BOTHHANDS));
        
         unusedTreasures.add(new Treasure("Insecticida",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Escopeta de 3 cañones",4,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Garabato místico",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("La rebeca metálica",2,TreasureKind.ARMOR));
         
         unusedTreasures.add(new Treasure("Lanzallamas",4,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Necromicón",1,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Necronomicón",5,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Linterna a 2 manos",3,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Necrognomicón",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Necrotelecom",2,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("Mazo de los antiguos",3,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Necroplayboycón",3,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Porra preternatural",2,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Shogulador",1,TreasureKind.BOTHHANDS));
         
         unusedTreasures.add(new Treasure("Varita de atizamiento",3,TreasureKind.ONEHAND));
         
         unusedTreasures.add(new Treasure("Tentáculo de pega",2,TreasureKind.HELMET));
         
         unusedTreasures.add(new Treasure("Zapato deja-amigos",1,TreasureKind.SHOES));
    }
    
    /**
     * Método initMonsterCarDeck()
     */
    private void initMonsterCardDeck(){
       BadConsequence badConsequence ;  // mal royo
       Prize prize; // buen royo
      
       ArrayList<TreasureKind> tVisible = new ArrayList(); // Lista de TreasureKind Visible
       ArrayList<TreasureKind> tHidden = new ArrayList(); // Lista de TresaureKind Oculto
       
       // 3 Byakhees de bonanza
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ARMOR));
       tHidden = new ArrayList(Arrays.asList(TreasureKind.ARMOR));
       badConsequence = new BadConsequence("Pierdes tu armadura visible y otra oculta"
               + "",0,tVisible,tHidden);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("3 Byakhees de bonanza ",8,badConsequence,prize));
       
       // Chibithulhu 
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.HELMET));
       tHidden = new ArrayList();
       badConsequence = new BadConsequence("Embobados con el lindo primigenio te descartas de tu"
               + "casco visible",0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Chibithulhu",2,badConsequence,prize));
       
       // El sopor de Dunwich
       tVisible = new ArrayList(Arrays.asList(TreasureKind.SHOES));
       tHidden = new ArrayList();
       badConsequence = new BadConsequence("El primordial bostezo contagioso.Pierdes el calzado visible"
                ,0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El sopor de Dunwich",2,badConsequence,prize));       
       
       
        // Ángeles de la noche ibicenca
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ONEHAND));
       tHidden = new ArrayList(Arrays.asList(TreasureKind.ONEHAND));
       badConsequence = new BadConsequence("Te atrapan para llevarte de fiesta "
               + "y te dejan caer en mitadl del vuelo. Descarta 1 mano visible "
               + "y 1 mano oculta",0,tVisible,tHidden);
       prize = new Prize(4,1);
       unusedMonsters.add(new Monster("Ángeles de la noche ibicenca",14,badConsequence,prize));
       
       //El gorrón en el umbral  
       
       badConsequence = new BadConsequence("Pierdes todos tus tesoros visibles",0,10,0);
       prize = new Prize(3,1);       
       unusedMonsters.add(new Monster("El gorrón en el umbral",10,badConsequence,prize));
       
       // H.P. Munchcraft
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ARMOR));
       tHidden = new ArrayList();
       badConsequence = new BadConsequence("Pierdes la armadura visible",0,tVisible,tHidden);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("H.P Munchcraft",6,badConsequence,prize));
         
       
       // Bichgooth
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ARMOR));
       tHidden = new ArrayList();
       badConsequence = new BadConsequence("Sientes bichos bajo la ropa.Descarta la armadura visible"
               ,0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Bichgooth",2,badConsequence,prize));
       
       
        
       // El rey de rosa
       badConsequence = new BadConsequence("Pierdes 5 niveles y 3 tesoros visibles",5,3,0);
       prize = new Prize(4,2);
       unusedMonsters.add(new Monster("El rey de rosa",13,badConsequence,prize));
       
        //La que redacta en las tinieblas
       
       badConsequence = new BadConsequence("Toses los pulmones y "
               + "pierdes 2 niveles",2,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("La que redacta en las tinieblas",2,badConsequence,prize));
       
       
       // Los hondos
       
       badConsequence = new BadConsequence("Estos monstrous resultan"
               + "bastante superficiales y te aburren mortalmente."
               + "Estas muerto",true);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Los hondos",8,badConsequence,prize));
       
       
       // Semillas Cthulhu
       
       badConsequence = new BadConsequence("Pierdes 2 niveles y 2 tesoros "
               + "ocultos",2,0,2);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Semillas Cthulhu",4,badConsequence,prize));
       
       // Dameargo
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ONEHAND));
       tHidden = new ArrayList();
       prize = new Prize(2,1);
       badConsequence = new BadConsequence("Te intentas escaquear. Pierdes"
               + "una mano visible",0,tVisible,tHidden);
       
       unusedMonsters.add(new Monster("Dameargo",1,badConsequence,prize));
       
       // Pollipólipo volante
       
       badConsequence = new BadConsequence("Da mucho asquito. Pierdes"
               + "3 niveles",3,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Pollipólipo volante",3,badConsequence,prize));
       
       // Yskhtihyssg-Goth
       
       badConsequence = new BadConsequence("No le hace gracia que pronuncien mal "
               + "su nombre. Estas muerto",true);
       prize = new Prize(3,1);
       unusedMonsters.add(new Monster("Yskhtihyssg-Goth",12,badConsequence,prize));
       
       // Familia feliz
       
       badConsequence = new BadConsequence("La familia te atrapa.Estás muerto"
               + "",true);
       prize = new Prize(4,1);
       unusedMonsters.add(new Monster("Familia Feliz",1,badConsequence,prize));
       
       
       //Roboggoth
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS));
       tHidden = new ArrayList();
       prize = new Prize(2,1);
       badConsequence = new BadConsequence("La quinta directiva primaria te obliga"
               + " a perder 2 niveles y un tesoro, 2 manos visibles",2,tVisible,tHidden);
       unusedMonsters.add(new Monster("Roboggoth",8,badConsequence,prize));
       
       //El espia
       
       tVisible = new ArrayList(Arrays.asList(TreasureKind.HELMET));
       tHidden = new ArrayList();
       badConsequence = new BadConsequence("Te asusta en la noche. Pierdes un "
               + "casco visible",0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El espia",5,badConsequence,prize));
       
       // El Lenguas
       
       badConsequence = new BadConsequence("Menudo susto te llevas."
               + " Pierdes 2 niveles y 5 tesoros visibles",2,5,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El Lenguas",20,badConsequence,prize));
       
       // Bicéfalo
       tVisible = new ArrayList(Arrays.asList(TreasureKind.ONEHAND));
       tHidden = new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS));;
       badConsequence = new BadConsequence("Te faltan manos para tanta cabeza. "
               + "Pierdes 3 niveles y tus tesoros visibles de las manos",3,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Bicéfalo",20,badConsequence,prize));
    }
    
    /**
     * Método shuffleTreasure()
     */
    private void shuffleTreasures(){
         Collections.shuffle(unusedTreasures);
    }
    
       
    /**
     * Método shuffleMonsters()
     */
    //Si usamos Collentions.shuffle(nombreaArray) las desordena tambien
    private void shuffleMonsters(){
        Collections.shuffle(unusedMonsters);
    }
    /**
     * Método getInstance()
     * @return 
     */
    public static CardDealer getInstance(){
        return instance;
    }
    
    /**
     * Método nextTresure()
     * @return 
     */
    public Treasure nextTreasure(){
        return null; //cambiar
    }
    
    /**
     * Método nextMonster()
     * @return 
     */
    public Monster nextMonster(){
        return null; //cambiar
    }
    
    /**
     * Método giveTreasureBack(Treasure t)
     * @param t 
     */
    public void giveTreasureBack(Treasure t){
       int j=0;
        
        usedTreasures.add(t);
              
       
    }
    
    /**
     * Método giveMonsterBack(Monster m)
     * @param m 
     */
    public void giveMonsterBack(Monster m){
        int j=0;
        
        usedMonsters.add(m);
               
        //unusedMonsters.remove(m);//si le metemos el ojeto lo busca y lo quita, si el objeto no existe no hace nada
    }
    
    /**
     * Método initCards()
     */
    public void initCards(){
        
    }
}

File: Juan/P2/Ruby/Napakalaki/lib/CardDealer.rb
require_relative 'Treasure.rb'
require_relative 'Monster.rb'
require_relative 'Prize.rb'
require_relative 'BadConsequence.rb'

include Singleton

class CardDealer
    
    #@@instance = nil
    private
    #/*Constructor/
    def initialize()
        @unusedMonsters=Array.new
        @usedMonsters=Array.new
        @unusedTreasures=Array.new
        @usedTreasures=Array.new
    end
    
    attr_reader :unusedMonster
    
    #/*Iniciar cartas tesoros/
    def initTreasureCardDeck
        
#        def initialize(n,b,t)
#        @name=n
#        @bonus=b
#        @treasure=t
#    end
    
        @unusedTreasures << Treasure.new('¡Si mi amo!', 4, TreasureKind::ARMOR)
        
        @unusedTreasures << Treasure.new('Botas de investigacion', 3, TreasureKind::SHOES)
        
        @unusedTreasures << Treasure.new('Capucha de Cthulhu', 3, TreasureKind::HELMET)
        
        @unusedTreasures << Treasure.new('A Prueba de babas', 2, TreasureKind::ARMOR)
        
        @unusedTreasures << Treasure.new('Botas de lluvia ácida',1 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Casco minero',2 , TreasureKind::HELMET)
        
        @unusedTreasures << Treasure.new('Ametralladora Thompson',4 , TreasureKind::BOTHHANS)
        
        @unusedTreasures << Treasure.new('Camiseta de UGR',1 , TreasureKind::ARMOR)
        
        @unusedTreasures << Treasure.new('Clavo de rail ferroviario',3 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Cuchillo de shusi arcano',2 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Fez alopodo', 3, TreasureKind::HELMET)
        
        @unusedTreasures << Treasure.new('Hacha prehistorica',3 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('El aparato del profesor Tesla',4 , TreasureKind::ARMOR)
        
        @unusedTreasures << Treasure.new('Gaita',4 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Insecticida',2 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Escopeta de tres cañones',4 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Garabato mistico',2 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('La rebeca metalica',2 , TreasureKind::ARMOR)
        
        @unusedTreasures << Treasure.new('Lanzallamas',4 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Necromicon',1 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Necronomicon',5 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Linterna a dos manos',3 , TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Necrognomicon',1 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Necrotelecom',2 , TreasureKind::HELMET)
        
        @unusedTreasures << Treasure.new('Mazo de los antiguos', 3, TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Necroplayboycon',3, TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Porra preternatural',2 , TreasureKind::ONEHAND)
        
        @unusedTreasures << Treasure.new('Shogulador', 1, TreasureKind::BOTHHANDS)
        
        @unusedTreasures << Treasure.new('Varita de atizamiento',3 , TreasureKind::ONEHAND)

        @unusedTreasures << Treasure.new('Tentaculo de pega',2 , TreasureKind::HELMET)
        
        @unusedTreasures << Treasure.new('Zapato deja-amigos',1 , TreasureKind::SHOES)
        
        
    end
    
    #/*Iniciar cartas monstruos/
    def initMonsterCardDeck
       

        # 3 Byakhees de bonanza

        price = Prize.new(2,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Pierdes tu armadura visible y otra oculta',0,[TreasureKind::ARMOR],[TreasureKind::ARMOR])
        @unusedMonsters << Monster.new('3 Byakhees de bonanza',8,badconsequence,price)

        # Chibithulhu

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Embobados don el lindo primigenio te descartas de tu casco visible',0,[TreasureKind::HELMET],[])
        @unusedMonsters << Monster.new('Chibithulhu',2,badconsequence,price)

        # El sopor de Dunwich

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('El primordial bostezo contagioso. Pierdes el calzado visible',0,[TreasureKind::SHOES],[])
        @unusedMonsters << Monster.new('El sopor de Dunwich',2,badconsequence,price)


        # Ángeles de la noche ibicenca

        price = Prize.new(4,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y  1 mano oculta',0 ,[TreasureKind::ONEHAND],[TreasureKind::ONEHAND])
        @unusedMonsters << Monster.new('Angeles de la noche ibicenca',14,badconsequence,price)

        # El gorrón en el umbral

        price = Prize.new(3,1)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes todos tus tesoros visibles',0,BadConsequence.MAXTREASURES,0)
        @unusedMonsters << Monster.new('El gorron en el umbral',10,badconsequence,price)

        BadConsequence.
        # H.P. Munchcraft

        price = Prize.new(2,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Pierdes la armadura visible',0,[TreasureKind::ARMOR],[])
        @unusedMonsters << Monster.new('H.P. Munchcraft',6,badconsequence,price)

        # Bichgooth

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Sientes bichos bajo la ropa. Descarta la armadura visible',0,[TreasureKind::ARMOR] , [] )
        @unusedMonsters << Monster.new('Bichgooth',2,badconsequence,price)


        # El rey de rosa

        price = Prize.new(4,2)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes 5 niveles y 3 tesoros visibles', 5, 3, 0)
        @unusedMonsters << Monster.new('El rey de rosa',13,badconsequence,price)

        # La que redacta en las tinieblas

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Toses los pulmones y pierdes 2 niveles',2,0,0)
        @unusedMonsters << Monster.new('La que redacta en las tinieblas',2,badconsequence,price)

        # Los hondos

        price = Prize.new(2,1)
        badconsequence = BadConsequence.newDeath("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto")
        @unusedMonsters << Monster.new('Los hondos',8,badconsequence,price)

        # Semillas Cthulhu

        price = Prize.new(2,1)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Pierdes 2 niveles y 2 tesoros ocultos',2,0,2)
        @unusedMonsters << Monster.new('Semillas Cthulhu',4,badconsequence,price)

        # Dameargo

        price = Prize.new(2,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Te intentas escaquear. Pierdes una mano visible',0,[TreasureKind::ONEHAND],[])
        @unusedMonsters << Monster.new('Dameargo',1,badconsequence,price)

        # Pollipólipo volante

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Da mucho asquito. Pierdes 3 niveles.',3,0,0)
        @unusedMonsters << Monster.new('Pollipolipo volante',3,badconsequence,price)

        # Yskhtihyssg-Goth

        price = Prize.new(3,1)
        badconsequence = BadConsequence.newDeath('No le hace gracia que pronuncien mal su nombre. Estas muerto')
        @unusedMonsters << Monster.new('Yskhtihyssg-Goth',12,badconsequence,price)

        # Familia feliz

        price = Prize.new(4,1)
        badconsequence = BadConsequence.newDeath("La familia te atrapa. Estas muerto")
        @unusedMonsters << Monster.new('Familia feliz',1,badconsequence, price)

        # Roboggoth
        price = Prize.new(2,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('La quinta directeiva primaria te obliga a perder 2 niveles y un tesoro, 2 manos visibles.',2,[TreasureKind::BOTHHANDS],[])
        @unusedMonsters << Monster.new('Roboggoth',8,badconsequence,price)

        # El espía

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Te asusta en la noche. Pierdes un casco visible',0,[TreasureKind::HELMET],[])
        @unusedMonsters << Monster.new('El espia',5,badconsequence,price)

        # El Lenguas

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelNumberOfTreasures('Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles',2, 5,0)
        @unusedMonsters << Monster.new('El Lenguas',20,badconsequence,price)

        # Bicéfalo

        price = Prize.new(1,1)
        badconsequence = BadConsequence.newLevelSpecificTreasures('Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos',3,[TreasureKind::BOTHHANDS,TreasureKind::ONEHAND,TreasureKind::ONEHAND],[])
        @unusedMonsters << Monster.new('Bicefalo',20,badconsequence,price)
    end
    
    #/*Barajar tesoros/
    def shuffleTreasures
        @unusedMonsters.shuffle!
    end
    
    #/*Barajar Monstruos/
    def shuffleMonsters
        @unusedTreasures.shuffle!
    end
    
    public
    #/*Devuelve instancia de barajador/
#    def self.getInstance
#       
#    end
    
    #/*siguiente tesoro/
    def nextTreasure
        
    end
    
    #/*siguiente monstruo/
    def nextMonster
        
    end
    
    #/*devuelve tesoro/
    def giveTreasureBack(t)
        @usedTreasures << t
    end
    
    #/*Devuelve monstruo/
    def giveMonsterBack(m)
        @usedMonsters << m
    end
    
    #/*Inicializar cartas/
    def initCards
        
    end
    
end